tests/extmod/machine_timer.py

```diff
@@ -0,0 +1,51 @@
+import sys
+
+try:
+    from machine import Timer
+    from time import sleep_ms
+except:
+    print("SKIP")
+    raise SystemExit
+
+if sys.platform in ("nrf"):
+    # The nrf port doesn't support selection of hard and soft callbacks,
+    # and only allows Timer(period=N), not Timer(freq=N).
+    print("SKIP")
+    raise SystemExit
+elif sys.platform in ("esp32", "esp8266"):
+    # Test with a hardware timer as software timers aren't implemented
+    # on the esp32 and esp8266 ports. Timer ID 0 is guaranteed to exist.
+    timer_id = 0
+else:
+    timer_id = -1
```

Comment:
It would be good to test with both hardware and software timers if possible on platforms that support it?

Reply:
Yes, it'd be great to make this more general to cover both on platforms that support both. Perhaps I can
```python
try:
  t = machine.Timer(0)
  # hardware timers available; test them
except:
  # hardware timers not available
```
and ditto for `Timer(-1)` for the soft timers. Is there anywhere where timer ID 0 is available but unsafe to use?

Reply:
As you suggested, I've added an extra test to cover hardware timers on the platforms that support them. I think `timer_id = 0` is the only choice that's guaranteed to be available on esp32 and esp8266, whereas `timer_id = 1` is a safe choice on stm32.